app/gui/view/graph-editor/src/lib.rs

```diff
@@ -3652,3 +3649,170 @@ impl display::Object for GraphEditor {
+        graph_editor.model.nodes.select(node_1_id);
+
+        // Adding second node.
+        let (_, node_2) = graph_editor.add_node_by(&add_node);
+        graph_editor.assert(Case { node_source: Some(node_1_id), should_edit: true });
+        assert_eq!(graph_editor.nodes().len(), 2);
+
+        // Second node is below the first.
+        let node_2_pos = node_2.position();
+        assert!(node_2_pos.y < node_1_pos.y);
+    }
+
+    #[test]
+    fn test_adding_node_by_dropping_edge() {
+        let (app, graph_editor) = init();
+        assert_eq!(graph_editor.model.nodes.all.len(), 0);
+        // Adding a new node.
+        let (node_1_id, node_1) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        // Creating edge.
+        let port = node_1.model.output_port().expect("No output port");
+        port.events.mouse_down.emit(());
+        assert_eq!(graph_editor.edges().len(), 1);
+        //node_1.model.output.test_port_press();
+        // Dropping edge.
+        let mouse = &app.display.default_scene.mouse;
+        let click_pos = Vector2(300.0, 300.0);
+        mouse.frp.position.emit(click_pos);
+        let click_on_background = |_: &GraphEditor| mouse.click_on_background();
+        let (_, node_2) = graph_editor.add_node_by(&click_on_background);
+        graph_editor.assert(Case { node_source: Some(node_1_id), should_edit: true });
+        let node_pos = node_2.position();
+        assert_eq!(node_pos.xy(), click_pos);
+    }
+
+    #[test]
+    fn test_connecting_two_nodes() {
+        let (_, ref graph_editor) = init();
+        let edges = graph_editor.edges();
+        assert!(graph_editor.nodes().is_empty());
+        assert!(edges.is_empty());
+        // Adding two nodes.
+        let (node_id_1, node_1) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        let (node_id_2, node_2) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        // Creating edge.
+        let port = node_1.model.output_port().expect("No output port");
+        port.events.mouse_down.emit(());
+        let edge_id = graph_editor.on_edge_add.value();
+        let edge = edges.get_cloned_ref(&edge_id).expect("Edge was not added");
+        assert_eq!(edge.source().map(|e| e.node_id), Some(node_id_1));
+        assert_eq!(edge.target().map(|e| e.node_id), None);
+        assert_eq!(edges.len(), 1);
+        // Connecting edge.
+        // We need to enable ports. Normally it is done by hovering the node.
+        node_2.model.input.frp.set_ports_active(true, None);
+        let port = node_2.model.input_port().expect("No input port");
+        port.hover.events.mouse_down.emit(());
+        assert_eq!(edge.source().map(|e| e.node_id), Some(node_id_1));
+        assert_eq!(edge.target().map(|e| e.node_id), Some(node_id_2));
+    }
+
+
+    // === Test utilities ===
+
+    struct Case {
+        node_source: Option<NodeId>,
+        should_edit: bool,
+    }
```

Comment:
Add docs. In what type of tests it is used?

Reply:
Done